### PR TITLE
Use GenISA to lower AddressPayloadRead (when env var TRITON_FORCE_GEN…

### PR DESCRIPTION
**Pull request by etiotto**

### CODE REVIEW

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 1]
```diff
@@ -387,9 +387,97 @@ createBlock2DReadWithAddressPayloadUpdate(TritonGEN::Matrix2DBlockLoadOp op,
...
+      fnName += "i16.p0i8";
+      break;
+    case 32:
+      fnName += "i32.p0i8";
+      break;
+    default:
+      llvm::errs() << "vecType.getElementType().getIntOrFloatBitWidth(): "
+                   << vecType.getElementType().getIntOrFloatBitWidth() << "\n";
+      llvm_unreachable("unhandled element size for LSC2DBlockReadAddrPayload");
+    }
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:433]: can we use `getGenISATypeMangling`?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 2]
```diff
@@ -387,9 +387,97 @@ createBlock2DReadWithAddressPayloadUpdate(TritonGEN::Matrix2DBlockLoadOp op,
...
+
+    Value zero = i32_val(0);
+    SmallVector<Type> argTypes{ptr.getType(), i32_ty, i32_ty, i32_ty, i32_ty,
+                               i32_ty,        i32_ty, i1_ty,  i1_ty,  i32_ty};
+
+    Value x = zero;
+    Value y = zero;
+    auto elemSize =
+        rewriter.create<LLVM::ConstantOp>(loc, i32_ty, op.getElemSizeInBits());
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:442]: can we use utilities in utility.h? like `i32_val(op.getElemSizeInBits())`

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 3]
```diff
@@ -387,9 +387,97 @@ createBlock2DReadWithAddressPayloadUpdate(TritonGEN::Matrix2DBlockLoadOp op,
                                     attrs);
   };
 
+  auto createBlock2DReadGenIsa = [&](Value ptr,
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:390]: [nit]
```suggestion
  auto createBlock2DReadGenISA = [&](Value ptr,
```

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 4]
```diff
@@ -387,9 +387,59 @@ createBlock2DReadWithAddressPayloadUpdate(TritonGEN::Matrix2DBlockLoadOp op,
                                     attrs);
   };
 
+  auto createBlock2DReadGenISA = [&](Value ptr,
+                                     TritonGEN::Matrix2DBlockLoadOp op) {
+    assert(isa<LLVM::LLVMPointerType>(ptr.getType()) &&
+           "Expecting a pointer type");
+
+    llvm::LLVMContext llvmContext;
+    LLVM::TypeToLLVMIRTranslator typeTranslator(llvmContext);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:396]: not used.
```suggestion
```

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 5]
```diff
@@ -387,9 +387,59 @@ createBlock2DReadWithAddressPayloadUpdate(TritonGEN::Matrix2DBlockLoadOp op,
...
+    assert(isa<VectorType>(resType) && "Expecting a vector type");
+    auto vecType = cast<VectorType>(resType);
+    assert(vecType.getShape().size() == 1 && "Expecting a 1D vector");
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:399]: [optional]
```suggestion
    auto vecType = dyn_cast<VectorType>(resType);
    assert(vecType && vecType.getShape().size() == 1 && "Expecting a 1D vector");
```

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 6]
```diff
@@ -387,9 +387,59 @@ createBlock2DReadWithAddressPayloadUpdate(TritonGEN::Matrix2DBlockLoadOp op,
...
+    SmallVector<Type> argTypes{ptr.getType(), i32_ty, i32_ty, i32_ty, i32_ty,
+                               i32_ty,        i32_ty, i1_ty,  i1_ty,  i32_ty};
+
+    Value x = zero, y = zero;
+    Value elemSize = i32_val(op.getElemSizeInBits());
+    Value tileWidth = i32_val(op.getTileWidth());
+    Value tileHeight = i32_val(op.getTileHeight());
+    Value vBlocks = i32_val(op.getVBlocks());
+    Value useTranspose = i1_val(op.getTranspose());
+    Value vnniTransform = i1_val(op.getVnniTransform());
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:414]: I would inline all these variables, as there is no extra info convey with them, and they are not used more than once.

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 4 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:396]: removed

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 5 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:399]: ok

**Pull request merged**